gh-107089: Improve Shelf.clear method performance

jtcave opened the pull request:
`Shelf` inherits its `clear` method from `MutableMapping`, but the implementation is poorly suited for that class. The `clear` mix-in is implemented by calling `popitem` in a loop. Each call to `popitem` constructs a new iterator over the shelf, which is a generator that does byte-to-str conversion on the keys in the `dbm` object. Unfortunately, `dbm` objects are non-iterable, and their `keys` method simply returns a list of all keys. Since each `popitem` call performs a full scan of the key space, the clear method ends up having O(n²) performance (assuming the delete and read operations amortize to O(1)).

By having the shelf just iterate over the list of dbm keys once, this is avoided.

<!-- gh-issue-number: gh-107089 -->
* Issue: gh-107089
<!-- /gh-issue-number -->

jtcave committed:
gh-107089: Improve Shelf.clear method performance

The clear method used to be implemented by inheriting a mix-in from the
MutableMapping ABC. It was a poor fit for shelves, and a better
implementation is now in place

corona10 commented on Lib/shelve.py:
```diff
@@ -171,6 +171,25 @@ def sync(self):
+                # that way.
+                keys = list(keys)
+            for k in keys:
+                del self.dict[k]
```
By https://github.com/python/cpython/issues/107122, now dbm/gdbm have the clear method, is this fallback implementation still needed?

jtcave replied:
It's still necessary for `dbm.dumb`.

corona10 replied:
`dbm.dumb.open` return a `_Database` and `_Database` is child class of `collections.abc.MutableMapping`
https://github.com/python/cpython/blob/2cf99026d6320f38937257da1ab014fc873a11a6/Lib/_collections_abc.py#L954-L973

So It has a clear method.

````
Python 3.13.0a0 (heads/main:2cf99026d6, Jul 24 2023, 20:24:49) [Clang 14.0.0 (clang-1400.0.29.202)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from dbm import dumb
>>> a = dumb.open('foo', 'c')
>>> type(a)
<class 'dbm.dumb._Database'>
>>> a['foo'] = 'bar'
>>> len(a)
1
>>> a.clear()
>>> len(a)
0
>>>
````

corona10 replied:
If you want to point out the performance issue of `dbm.dumb`, we can add the specialized method like ndbm and gdbm.

jtcave replied:
The fallback code isn't any faster than the clear method in `dbm.dumb`. Looking at the code, that module doesn't have the quadratic flaw of the old gdbm/ndbm implementations (its `__iter__` just returns an iterator for a dict), so the issue doesn't apply.

I've updated the PR to just call the clear method on the backing object unconditionally, as well as to move the method to `DbfilenameShelf`.

jtcave committed:
gh-107089: call dbm.clear method

Because gh-107089 is peculiar to implementation details of dbm objects,
it would be less disruptive to implement it in the DbfilenameShelf
class, which is used for calls to shelve.open. Since it is known that
the backing object is specifically one of the dbm objects, its clear
method (see gh-107122) can be used with no fallback code.

jtcave committed:
fix whitespace

corona10 commented on Lib/shelve.py:
```diff
@@ -171,7 +171,6 @@ def sync(self):
             self.dict.sync()
 
-
 class BsdDbShelf(Shelf):
```
```suggestion

class BsdDbShelf(Shelf):
```

jtcave replied:
applied ✔️

jtcave committed:
fix whitespace

corona10 reviewed (approved):
lgtm!
